File: photomap-core-api/photo-api/src/main/java/ua/in/photomap/photoapi/validator/CommonPhotoValidator.java
package ua.in.photomap.photoapi.validator;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.util.ObjectUtils;
import org.springframework.web.multipart.MultipartFile;
import ua.in.photomap.common.rest.toolkit.exception.ValidationException;
import ua.in.photomap.photoapi.dto.PhotoMetadataDTO;
import ua.in.photomap.photoapi.service.LicenseService;

import java.util.Arrays;

@Component
@RequiredArgsConstructor
public class CommonPhotoValidator {
    private final LicenseService licenseService;

    private static final String[] ALLOWED_CONTENT_TYPES = {
            "image/jpeg"
    };

    @Value("${photomap.file.max-size-in-mb}")
    private Long maxFileSizeInMb;

    public void validateMetadata(PhotoMetadataDTO metadata) {
        if (ObjectUtils.isEmpty(metadata)) {
            throw new ValidationException("Metadata is required");
        } else if (metadata.getLicenseId() == null) {
            throw new ValidationException("License is required");
        } else if (licenseService.findById(metadata.getLicenseId()).isEmpty()) {
            throw new ValidationException("License with id " + metadata.getLicenseId() + " does not exist");
        } else if (metadata.getYearRange() == null || metadata.getYearRange().getStart() == null
                || metadata.getYearRange().getEnd() == null) {
            throw new ValidationException("Year range is required");
        } else if (metadata.getTags() == null || metadata.getTags().isEmpty()
                || metadata.getTags().stream().anyMatch(ObjectUtils::isEmpty)) {
            throw new ValidationException("Tags cannot be empty");
        } else if (metadata.getCoordinates() == null || metadata.getCoordinates().getLatitude() == null
                || metadata.getCoordinates().getLongitude() == null) {
            throw new ValidationException("Coordinates are required");
        } else if (metadata.getCoordinates().getIsApproximate() == null) {
            throw new ValidationException("Coordinates approximation is required");
        }
    }

    public void validateFile(MultipartFile file) {
        if (ObjectUtils.isEmpty(file)) {
            throw new ValidationException("File is empty.");
        }
        if (!Arrays.asList(ALLOWED_CONTENT_TYPES).contains(file.getContentType())) {
            throw new ValidationException("Invalid file type. Only images are allowed.");
        }

        if (file.getSize() > getMaxFileSizeInBytes()) {
            throw new ValidationException("File size exceeds the allowed limit of " + maxFileSizeInMb + "MB");
        }
    }

    private Long getMaxFileSizeInBytes() {
        return maxFileSizeInMb * 1024 * 1024;
    }

}

File: photomap-core-api/photo-api-python/validators/common_photo_validator.py
from werkzeug.datastructures import FileStorage
from services.license_service import LicenseService
from dto.photo_metadata_dto import PhotoMetadataDTO
from exceptions.common_exceptions import ValidationException
from config import Config
import logging

logger = logging.getLogger(__name__)

class CommonPhotoValidator:

    ALLOWED_CONTENT_TYPES = ['image/jpeg']
    
    def __init__(self, license_service: LicenseService):
        self.license_service = license_service
    
    def validate_metadata(self, metadata: PhotoMetadataDTO):
        if not metadata:
            raise ValidationException("Metadata is required")

        if metadata.licenseId is None:
            raise ValidationException("License is required")
        
        if not self.license_service.find_by_id(metadata.licenseId):
            raise ValidationException(f"License with id {metadata.licenseId} does not exist")

        if metadata.yearRange is None or metadata.yearRange.start is None or metadata.yearRange.end is None:
            raise ValidationException("Year range is required")

        if metadata.tags is None or len(metadata.tags) == 0:
            raise ValidationException("Tags cannot be empty")
        
        for tag in metadata.tags:
            if not tag or tag.strip() == '':
                raise ValidationException("Tags cannot be empty")

        if metadata.coordinates is None:
            raise ValidationException("Coordinates are required")
        
        if metadata.coordinates.latitude is None or metadata.coordinates.longitude is None:
            raise ValidationException("Coordinates are required")
        
        if metadata.coordinates.isApproximate is None:
            raise ValidationException("Coordinates approximation is required")
    
    def validate_file(self, file: FileStorage):
        if not file or not file.filename:
            raise ValidationException("File is empty.")

        if file.content_type not in self.ALLOWED_CONTENT_TYPES:
            raise ValidationException("Invalid file type. Only images are allowed.")

        file.seek(0, 2)  # Seek to end
        file_size = file.tell()
        file.seek(0)  # Reset position
        
        max_size_bytes = Config.MAX_FILE_SIZE_MB * 1024 * 1024
        if file_size > max_size_bytes:
            raise ValidationException(f"File size exceeds the allowed limit of {Config.MAX_FILE_SIZE_MB}MB") 